server/src/game/module.js

```diff
@@ -42,47 +45,133 @@ class GameModule {
 
   assignRoles(clients, sessionRoles) {
     clients.forEach((client, i) => this.players[i] = new Player(client.name, client.id, sessionRoles[i], i));
+    // guarantee order is maintained
+    Object.freeze(this.players);
   }
 
   assignMiddle(sessionRoles) {
-    this.middle = sessionRoles.slice(-3);
+    /** @type {Array<Card>} */
+    this.middle = sessionRoles.slice(-3).map((role) => new Card(role));
+    Object.freeze(this.middle);
   }
 
+  /** @returns {boolean} */
   get hasNextRole() { return this.sessionOrder.length > 0; }
 
+
+  /**
+   * @typedef {Object} NextRoleData
+   * @property {Array<Player>} playerTargets
+   * @property {Array<Object>} roleData
+   * @property {String} dialogue
+   */
+
+  /**
+   * @returns {NextRoleData}
+   */
   readyNextRole() {
     if(this.sessionOrder.length === 0) throw new Error("No more roles");
 
-    let nextRole;
+    let nextRole = true;
     // error handling in case `this.sessionOrder` contains other roles
-    while(!this.gameHasRole(nextRole = this.sessionOrder.shift()));
+    while(nextRole && !this.gameHasRole(nextRole = this.sessionOrder.shift()));
+    if(!nextRole) throw new Error("No more roles");
     console.log(`Readying next role: ${nextRole}`);
 
     let playerTargets = this.players.filter((player) => nextRole === player.originalRole);
 
     const dialogue = getDialogue(nextRole);
-    const roleData = getRoleData(nextRole);
+    const roleData = getRoleData(nextRole, playerTargets);
 
+    /** @type {String} */
     this.currentRole = nextRole;
-    this.roleActed = false;
+
+    if(isEffectivelyPassive(this.currentRole)) {
+      playerTargets.forEach((player) => this.playerAct(player));
+    }
 
     return { playerTargets, roleData, dialogue };
   }
 
+  /** @returns {boolean} */
   gameHasRole(role) {
     return this.players.filter((player) => player.originalRole === role).length > 0;
   }
 
-  playerAct(id, data) {
+  /**
+   * @param {Number} id 
+   * @returns {Player}
+   */
+  getPlayer(id) {
     const player = this.players.filter((el) => el.id === id)[0];
-    if(!player || player.originalRole !== this.currentRole || player.hasActed) return;
+    if(!player) throw new Error(`No player found with ID ${id}`);
+    return player;
+  }
 
-    console.log(`Player ${player.name} acting`);
+  playerAct(player, data) {
+    if(player.originalRole !== this.currentRole || player.hasActed || player.asleep) return;
 
-    // TODO: use `data` to manipulate roles, players, etc.
-    // TODO: check if player is asleep. If so, don't do anything
+    console.log(`Player ${player.name} acting`);
+    let responseData;
+    let fullyActed = true;
+
+    if(player.originalRole === roles.sydney) {
```

Comment:
That's smart. I just pushed a commit doing this.

Honestly, I think the way this function is written is pretty gross, I just couldn't think of a better way. A switch block isn't ideal because you can't declare new variables in case statements.